### PR TITLE
Fix diskann memory problem

### PR DESCRIPTION
**Pull request by hhy3**
Signed-off-by: zh Wang <zihao.wang@zilliz.com>

issue: #437 

This PR: 
1. Fixed the mixed use of malloc/delete and new/free.
2. Upgrade C++ standard to from C++14 to C++17 in DiskANN.

### CODE REVIEW

**Comment by liliu-z** [comment]
> Signed-off-by: zh Wang [zihao.wang@zilliz.com](mailto:zihao.wang@zilliz.com)
> 
> issue: #437
> 
> This PR fixed the mixed use of malloc/delete and new/free.
> 
> It's better to use std::any instead of boost::any, but it turns out that currently diskann doesn't support c++17. Maybe we should upgrade the standard to c++17.

What do you mean by doesn't support c++17. Knowhere is using C++17 right now.

File: knowhere/index/vector_index/IndexDiskANN.cpp [hunk 1]
```diff
@@ -373,8 +373,8 @@ IndexDiskANN<T>::Query(const DatasetPtr& dataset_ptr, const Config& config, cons
 
     GET_TENSOR_DATA_DIM(dataset_ptr);
     auto query = static_cast<const T*>(p_data);
-    auto p_id = static_cast<int64_t*>(malloc(sizeof(int64_t) * k * rows));
```
Comment (matchyc) [inline knowhere/index/vector_index/IndexDiskANN.cpp:376]: Check this response, maybe it's not a good idea?
https://github.com/milvus-io/knowhere/pull/298#discussion_r926881741

**Comment by hhy3** [comment]
> What do you mean by doesn't support c++17. Knowhere is using C++17 right now.

Knowhere is using C++17 but inside thirdparty/DiskANN it only supports C++14 now https://github.com/milvus-io/knowhere/blob/main/thirdparty/DiskANN/CMakeLists.txt#L26 and it can't compile when I use C++17 feature inside thirdparty/DiskANN

File: knowhere/index/vector_index/IndexDiskANN.cpp (hunk 1 above)
Comment (hhy3) [reply knowhere/index/vector_index/IndexDiskANN.cpp:376]: In #350 konwhere has replaced malloc/free with new/delete. So it may not be a problem.

**Pull request merged**